src/python/pants/engine/goal.py

```diff
@@ -0,0 +1,102 @@
+# coding=utf-8
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from contextlib import contextmanager
+
+from pants.cache.cache_setup import CacheSetup
+from pants.option.optionable import Optionable
+from pants.option.scope import ScopeInfo
+from pants.subsystem.subsystem_client_mixin import SubsystemClientMixin
+from pants.util.meta import AbstractClass, classproperty
+
+
+class Goal(SubsystemClientMixin, Optionable, AbstractClass):
+  """A CLI goal whch is implemented by a `@console_rule`.
+  
+  This abstract class should be subclassed and given a `Goal.name` that it will be referred to by
+  when invoked from the command line. The `Goal.name` also acts as the options_scope for the `Goal`.
+  """
+
+  # Subclasser-defined. See the class pydoc.
+  name = None
+
+  # If this Goal should have an associated deprecated instance of `CacheSetup` (which was implicitly
+  # required by all v1 Tasks), subclasses may set this to a valid deprecation version to create
+  # that association.
+  deprecated_cache_setup_removal_version = None
+
+  @classproperty
+  def options_scope(cls):
+    if not cls.name:
+      # TODO: Would it be unnecessarily magical to have `cls.__name__.lower()` always be the name?
+      raise AssertionError('{} must have a `Goal.name` defined.'.format(cls.__name__))
+    return cls.name
+
+  @classmethod
+  def subsystem_dependencies(cls):
+    # NB: `Goal` implements `SubsystemClientMixin` in order to allow v1 `Tasks` to depend on
+    # v2 Goals, and for `Goals` to declare a deprecated dependency on a `CacheSetup` instance for
+    # backwards compatibility purposes. But v2 Goals should _not_ have subsystem dependencies:
+    # instead, the @rules participating (transitively) in a Goal should directly declare
+    # Subsystem deps.
+    if cls.deprecated_cache_setup_removal_version:
```

Comment:
In cases like `list`, they were never actually useful. But immediately removing them would break cases where folks might have thought that they were useful, without notice. I agree that this is more a "letter of the deprecation law" situation, but I added this facility because `list` is unlikely to be the only case we encounter during goal porting, and there are a bunch of cases where people were more likely to make this mistake.